opm/parser/eclipse/Parser/ErrorGuard.hpp

```diff
@@ -0,0 +1,51 @@
+/*
+  Copyright 2019 Joakim Hove/datagr
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef ERROR_GUARD_HPP
+#define ERROR_GUARD_HPP
+
+#include <string>
+#include <vector>
+
+namespace Opm {
+
+class ErrorGuard {
+public:
+    void add_error(const std::string& errorKey, const std::string& msg);
+    void add_warning(const std::string& errorKey, const std::string &msg);
+
+    explicit operator bool() const { return !this->error_list.empty(); }
+
+    /*
+      Observe that this desctructor has a somewhat special semantics - it will
+      throw std::runtime_error if there are errors in the error_list.
+    */
+    ~ErrorGuard();
```

Comment:
The comment is not consistent with the implementation, which terminates rather than throwing. I think throwing would be a better solution, it would also be possible to test...

It also does not mention that the messages will be written to std::cerr.

Reply:
As I always - comments quickly go stale .... Will update.

Reply:
> I think throwing would be a better solution, it would also be possible to test...

I agree on that - but there are some low level things about throwing in the destructor which are above my C++ pay grade; at least my naive attempts at this failed - I'll give it another shot.

Reply:
It's true that it is not unproblematic. If the destructor is called as a result of stack unwinding due to *another* exception being thrown, and it then throws itself, std::terminate() will end up being called anyway. But at least that is possible to control using try/catch etc.